### PR TITLE
[TP] Add an input resharding wrapper for TP and unit test for 2D + AC

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103334

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
Curious, why is it called `recompute_wrapper` - are we doing actual activation recomputation here, or just distributed / gather the activation?

**Comment by fduwjj** [comment]
@rohan-varma the original checkpointing name is easy to cause confuse because people usually confuse it with model checkpointing. And yes within this wrapper, we also do activation recomputing. I am open to better name, do you have any suggestions?

**Comment by awgu** [comment]
Which name to use has always been confusing because the original paper used "gradient checkpointing", but in reality you are checkpoint activations. At the same time, you are also recomputing activations, so "activation recomputation" is another valid name.

@fduwjj  @rohan-varma  My opinion is that we should stay consistent with what already exists in PyTorch, which is to still refer to it as "activation checkpointing" and use the term "checkpoint".

File: torch/distributed/tensor/parallel/recompute_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    CheckpointImpl,
+    CheckpointWrapper,
+)
+
+__all__ = [
+    "tp_checkpoint_wrapper",
+]
+
+
+class TPCheckpointWrapper(CheckpointWrapper):
```
Comment (albanD) [inline torch/distributed/tensor/parallel/recompute_wrapper.py:18]: I'm curious why you subclass from CheckpointWrapper here?
Why is this not an independent feature like AutoReshard() and then the caller is expected to do `AutoReshard(CheckpointWrapper(mod))` ?

That would allow you to separate the discussion about what this code does from the checkpoint code?

Comment (fduwjj) [reply torch/distributed/tensor/parallel/recompute_wrapper.py:18]: because we want to this wrapper to be compatible with 2D parallel, so that it can work with a module wrapped with FSDP.

Comment (wconstab) [reply torch/distributed/tensor/parallel/recompute_wrapper.py:18]: i dunno if you answered alban's question (technically) - couldn't you use Reshard(Checkpoint()) together with FSDP? I guess the question was about whether it needs to be a monolithic thing to work correctly, or if it can be 2 things that can compose

Comment (fduwjj) [reply torch/distributed/tensor/parallel/recompute_wrapper.py:18]: Is the Reshard here a self-defined autograd function or?

Comment (albanD) [reply torch/distributed/tensor/parallel/recompute_wrapper.py:18]: Reshard would be exactly the implementation you have here minus the call into the checkpoint Module.
So they can compose exactly the same way but you don't have to use Checkpoint when you use Reshard.

Comment (fduwjj) [reply torch/distributed/tensor/parallel/recompute_wrapper.py:18]: Oh, I see your point, maybe I can make it just a subclass of `nn.Module`. Thanks!

Comment (albanD) [reply torch/distributed/tensor/parallel/recompute_wrapper.py:18]: That sounds great! Thanks!

Comment (fduwjj) [reply torch/distributed/tensor/parallel/recompute_wrapper.py]: Addressed your feedback, LMK if you have other comments.

File: test/distributed/tensor/parallel/test_2d_parallel.py [hunk 2]
```diff
@@ -77,9 +97,16 @@ def init_model(model_parallel_size=TP_DEGREE, use_orig_params=False, fsdp_nested
...
+        model,
+        True,
+        twod_mesh,
+        fsdp_pg,
+        use_orig_params,
+        fsdp_nested,
+        recompute_activation,
     )
-    return model, fsdp_pg
+    tp_mesh = _create_1d_device_mesh(twod_mesh, 1)
```
Comment (kumpera) [inline test/distributed/tensor/parallel/test_2d_parallel.py:108]: nit: This sort of mesh manipulation should be an integral part of DeviceMesh itself.

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 3]
```diff
@@ -123,6 +126,14 @@ def test_mlp_megatron_e2e_w_tensor_parallel(self):
     def test_mlp_megatron_e2e_w_sequence_parallel(self):
         self._test_mlp_magatron_e2e(is_seq_parallel=True)
 
+    @with_comms
```
Comment (kumpera) [inline test/distributed/tensor/parallel/test_tp_examples.py:129]: This would be easy to test using parameterize and cover all combinations.

File: torch/distributed/tensor/parallel/input_reshard.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+from torch.distributed._tensor import DeviceMesh, DTensor, Replicate, Shard
+
+__all__ = [
+    "input_reshard_wrapper",
+]
+
+
+class InputReshard(torch.nn.Module):
+    """
+    An nn.Module that wraps another nn.Module with input resharding.
```
Comment (kumpera) [inline torch/distributed/tensor/parallel/input_reshard.py:16]: Can't this be done with fwd hooks instead of a wrapper?

File: torch/distributed/tensor/parallel/input_reshard.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+def input_reshard_wrapper(
+    module: torch.nn.Module,
+    tp_device_mesh: DeviceMesh,
+    input_reshard_dim: Optional[int] = None,
+) -> torch.nn.Module:
+    """
+    Wrap an nn.Module with input resharding so that we can shard
+    per the given `tp_device_mesh` and `input_reshard_dim` and restore the
+    input back when recomputing the activations in the backward. The reason
+    why we can do this is that for Tensor Parallel(TP), the input are same
```
Comment (kumpera) [inline torch/distributed/tensor/parallel/input_reshard.py:68]: We should think on a way to have this invariant checked when ppl use this through the composable API.

File: torch/distributed/tensor/parallel/input_reshard.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    across all TP ranks.
+
+    Args:
+        module (:class:`nn.Module`):
+            Module to be wrapped with input resharding.
+        tp_device_mesh (:class:`DeviceMesh`):
```
Comment (kumpera) [inline torch/distributed/tensor/parallel/input_reshard.py:74]: From looking at the code, it's not clear to me why we need to pass a DM.
Is this DM different from the one used by the DTensors in the module? If yes, what's the expected difference?

**Review by kumpera — CHANGES_REQUESTED** [review]

File: torch/distributed/tensor/parallel/input_reshard.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        mod: torch.nn.Module,
+        tp_device_mesh: DeviceMesh,
+        input_reshard_dim: Optional[int] = None,
+    ):
+        super(InputReshard, self).__init__()
+        self.inner_module = mod
+        self.mesh = tp_device_mesh
+        self.input_reshard_dim = input_reshard_dim
+
+    def named_parameters(
```
Comment (awgu) [inline torch/distributed/tensor/parallel/input_reshard.py:34]: I want to mention that doing this can be brittle for some cases since now iterating over `named_modules()` and constructing parameter names manually differs from `named_parameters()`. This is not necessarily common, but it can happen.

For example, nanoGPT's original optimizer construction did this (see the 'before'):
https://github.com/karpathy/nanoGPT/commit/7fe4a099ad2a4654f96a51c0736ecf347149c34c

File: test/distributed/tensor/parallel/test_2d_parallel.py (hunk 2 above)
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_2d_parallel.py:108]: Yes I totally agree but it's outside the scope of this PR. I can send a follow-up PR to address this or someone is already working on it?

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 4 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/input_reshard.py:16]: But within it, it's a context manager, is it possible we can put that inside a fwd hook? cc: @albanD

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 5 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/input_reshard.py:68]: This will always be a API and decided by users?

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 6 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/input_reshard.py:74]: Well, input is tensor essentially and we need DM to reshard input when input is not a DTensor, which is pretty common for TP right now. We add a pre-fwd hook to convert tensor to dtensor. And then this reshard is being triggered, our pre-forward hook has not been triggered yet.

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 7 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/input_reshard.py:34]: I see. I borrowed this idea from FSDP's AC wrapper. Hmm after reading composable API's AC, I think we might be able to make it a pre-forward hook.. let me try.

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 4 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/input_reshard.py]: NVM I made it as two hooks.

File: test/distributed/tensor/parallel/test_2d_parallel.py (hunk 2 above)
Comment (kumpera) [reply test/distributed/tensor/parallel/test_2d_parallel.py:108]: A follow up PR would be beautiful!

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 5 above)
Comment (kumpera) [reply torch/distributed/tensor/parallel/input_reshard.py:22]: Sure, the user can decide to use it or not. 
Enforcing the invariant that the input is replicates across `tp_device_mesh` is left as an exercise for the user.

As part of making 2D/TP prod ready, we need a way to automatically verify that invariant on behalf of our users to make it easy for them to find bugs.

For example.

Say we have a `enable_2d_debugging` function in composability. Could we check this by having the pre_forward hook hash the input tensors, all_gather the hashes and compare them? Would that be enough for us to ensure input_resharding will work as intended?

File: torch/distributed/tensor/parallel/input_reshard.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    """
+    cx: Optional[torch.autograd.graph.saved_tensors_hooks] = None
+
+    def input_reshard_forward_pre_hook(_: torch.nn.Module, _i: Tuple[Any, ...]) -> None:
+        saved_tensor_hooks = torch.autograd.graph.saved_tensors_hooks(
+            partial(_pack_hook_tp, tp_device_mesh, input_reshard_dim),
+            partial(_unpack_hook_tp, tp_device_mesh, input_reshard_dim),
+        )
+        saved_tensor_hooks.__enter__()
+        global cx
```
Comment (kumpera) [inline torch/distributed/tensor/parallel/input_reshard.py:47]: `cx` is not a global, shouldn't this use `nonlocal` instead?

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 6 above)
Comment (kumpera) [reply torch/distributed/tensor/parallel/input_reshard.py:28]: That makes perfect sense, thanks for explaining it.

**Review by kumpera — APPROVED** [review]
LGTM

File: torch/distributed/tensor/parallel/input_reshard.py (hunk 8 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/input_reshard.py:47]: oh, I see. Let me change it.
